### PR TITLE
[SPARK-11448][SQL] Skip caching part-files in driver side when configured to merge schema and respect summaries

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-11448

We now cache part-files, metadata, common metadata in ParquetRelation as currentLeafStatuses. However, when configured to merge schema and respect summaries, dataStatuses (`FileStatus` objects of all part-files) are not necessary anymore. We should skip them when caching in driver side.

### CODE REVIEW

**Comment by viirya** [comment]
cc @liancheng

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 1]
```diff
@@ -208,7 +208,15 @@ private[sql] class ParquetRelation(
   // Parquet data source always uses Catalyst internal representations.
   override val needConversion: Boolean = false
 
-  override def sizeInBytes: Long = metadataCache.dataStatuses.map(_.getLen).sum
+  override def sizeInBytes: Long =
+    if (shouldMergeSchemas && mergeRespectSummaries) {
+      // If we are going to merge schema and this relation is configured to
+      // respect summaries (i.e., skip part-files), we will assume that the size of
+      // this relation is large and can't be broadcasted.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: This assumption doesn't seem right to me... It's perfectly OK for a small Parquet dataset to require schema merging and to be configured to respect summaries.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Right. But compared to set a smaller `sizeInBytes`, a larger one seems better as wrongly broadcasting large relation should be worse?

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 2]
```diff
@@ -434,10 +439,7 @@ abstract class HadoopFsRelation private[sql](maybePartitionSpec: Option[Partitio
 
           logInfo(s"Listing $qualified on driver")
           Try(fs.listStatus(qualified)).getOrElse(Array.empty)
-        }.filterNot { status =>
-          val name = status.getPath.getName
-          name.toLowerCase == "_temporary" || name.startsWith(".")
-        }
+        }.filterNot(fileStatusFilter)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: In `ParquetRelation`, we inject cached `FileStatus`es into an anonymous extended `ParquetInputFormat`. I'd expect that disabling caching data files should cause test failures, not sure why no test failure occurred :( Either I missed something here or our test cases are rather incomplete.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I don't disable caching data but just filter out some `FileStatus`es.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I mean this disables caching `FileStatus`es of part-files, and this does influence `ParquetInputFormat.listStatus()`. I believe the reason why no test case was broken is that currently `ParquetInputFormat` only calls `listStatus()` for reading footers. But this is completely an implementation detail within Parquet itself and shouldn't be viewed as a stable assumption.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Yeah, I understand that this is a "safe" assumption. However, we don't need to do this assumption if `FileStatus`es of part-files are properly cached.

**Comment by liancheng** [comment]
Hey @viirya, as mentioned in my inline comments, I have two major concerns about this change:

1. It may introduce performance regression, i.e., in some cases small Parquet dataset can't be broadcasted.
2. It relies on an assumption that completely depends on Parquet internal implementation details

On the other hand, it's not clear that how much we gain from this change.

**Comment by liancheng** [comment]
Hey @viirya, are you still working on this?

**Comment by viirya** [comment]
@liancheng your suggestion? Close it first? I think this will not introduce much performance gain but only to reduce memory pressure on driver side for the case with too many part-files.

**Comment by liancheng** [comment]
Yeah, since there's not much gain and may even introduce regression in some cases. Lets close it. Thanks all the same!

**Pull request closed without merging**